### PR TITLE
[TBL] Sign in using the OAuth 2 service directly

### PR DESCRIPTION
**Pull request by jcox-dev**
Change ras-frontstage to directly call the Django OAuth 2 and Party services instead of proxying through ras-frontstage-api.

Notable changes include:
- creation of the `oauth_controller` and `party_controller` (from frontstage_api)
- new exception, `OAuth2Error`, for presenting detailed information to the user
- service URLs added as config
- moving from "DJANGO_" to "OAUTH_" variable names to be explicit

The API calls to the service adds overhead in the following ways:

- Development - Code needs to be changed in ras-frontstage and ras-frontstage-api.
- Debugging - Unnecessary hops makes debugging more difficult
- Code reviews - split over multiple repos

This is in an effort to deprecate frontstage-api by refactoring to use in-house controllers that replicate the implementation from fronstage-api thus reducing networking overhead.

NB: the jenkins jobs need their config updated, otherwise this'll fail spectacularly!

### CODE REVIEW

File: tests/app/mocked_services.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import json
+from frontstage import app
+
+
+with open('tests/test_data/party/party.json') as json_data:
+	party = json.load(json_data)
+
+
+url_oauth_token = f"{app.config['OAUTH_SERVICE_URL']}/api/v1/tokens/"
+url_get_party_email = f"{app.config['PARTY_SERVICE_URL']}/party-api/v1/respondents/email"
```
Comment (rhiGriff) [inline tests/app/mocked_services.py]: Missing new line at end of file 🎨

File: frontstage/controllers/oauth_controller.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        'client_id': app.config['OAUTH_CLIENT_ID'],
+        'client_secret': app.config['OAUTH_CLIENT_SECRET'],
+        'username': username,
+        'password': password,
+    }
+    headers = {
+        'Accept': 'application/json',
+        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
+    }
+    response = requests.post(url, headers=headers, auth=app.config['OAUTH_BASIC_AUTH'], data=data)
```
Comment (JamesGardiner) [inline frontstage/controllers/oauth_controller.py]: Are network level errors handled by the application, or should they be handled here?

File: frontstage/controllers/oauth_controller.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    except requests.exceptions.HTTPError:
+        if response.status_code == 401:
+            oauth2_error = json.loads(response.text).get('error').get('data', {}).get('detail')
+            logger.exception('Authentication error in oauth2 service', error=oauth2_error)
+            raise OAuth2Error(response, message=oauth2_error)
+        else:
+            logger.exception('Failed to retrieve OAuth2 token')
+        raise ApiError(response)
+
+    oauth2_token = json.loads(response.text)
```
Comment (JamesGardiner) [inline frontstage/controllers/oauth_controller.py]: Decode errors should be handled here.

File: frontstage/controllers/oauth_controller.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        'Accept': 'application/json',
+        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
+    }
+    response = requests.post(url, headers=headers, auth=app.config['OAUTH_BASIC_AUTH'], data=data)
+
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError:
+        if response.status_code == 401:
+            oauth2_error = json.loads(response.text).get('error').get('data', {}).get('detail')
```
Comment (JamesGardiner) [inline frontstage/controllers/oauth_controller.py]: Requests decoder is far better at guessing encodings than standard json.loads, so the `Response.json` method should always be preferred to this approach:

```Python
oauth2_error = response.json().get('error', {}).get('data', {}).get('detail')
```

File: frontstage/controllers/oauth_controller.py (hunk 3 above)
Comment (JamesGardiner) [inline frontstage/controllers/oauth_controller.py]: Same here, use the `response.json()` method.

File: frontstage/controllers/party_controller.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+from frontstage.exceptions.exceptions import ApiError
+
+
+logger = wrap_logger(logging.getLogger(__name__))
+
+
+def get_party_by_email(email):
+    logger.debug('Retrieving party')
+    url = f"{app.config['PARTY_SERVICE_URL']}/party-api/v1/respondents/email"
+    response = requests.get(url, json={"email": email}, auth=app.config['BASIC_AUTH'])
```
Comment (JamesGardiner) [inline frontstage/controllers/party_controller.py]: Same question here as before - where are network level errors handled?

File: frontstage/controllers/party_controller.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError:
+        logger.exception('Failed to retrieve party')
+        raise ApiError(response)
+
+    logger.debug('Successfully retrieved party')
+    return json.loads(response.text)
```
Comment (JamesGardiner) [inline frontstage/controllers/party_controller.py]: response.json()

File: frontstage/views/sign_in/sign_in.py [hunk 7]
```diff
@@ -30,49 +36,40 @@ def login():
...
-                logger.error('OAuth 2 server generated 401 which is not understood',
-                             oauth2error=error_message)
-                return render_template('sign-in/sign-in.html', form=form, data={"error": {"type": "failed"}})
-
-        if response.status_code != 200:
-            logger.error('Failed to sign in')
-            raise ApiError(response)
+            logger.error('OAuth 2 server generated 401 which is not understood',
+                         oauth2error=error_message)
+            return render_template('sign-in/sign-in.html', form=form, data={"error": {"type": "failed"}})
```
Comment (JamesGardiner) [inline frontstage/views/sign_in/sign_in.py]: An explicit else, while not required, would enhance the clarity of this if/else block.

File: tests/app/mocked_services.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import json
+from frontstage import app
```
Comment (JamesGardiner) [inline tests/app/mocked_services.py]: Blank line needed between different import sections

File: frontstage/controllers/oauth_controller.py (hunk 4 above)
Comment (jcox-dev) [reply frontstage/controllers/oauth_controller.py]: Lazy copy pasta  👍

File: frontstage/controllers/oauth_controller.py (hunk 2 above)
Comment (jcox-dev) [reply frontstage/controllers/oauth_controller.py]: Yup - [error_handlers.py](https://github.com/ONSdigital/ras-frontstage/blob/master/frontstage/error_handlers.py#L22).

Comment (jcox-dev) [reply frontstage/controllers/oauth_controller.py]: @JamesGardiner should be handled by the [error_handlers.py](https://github.com/ONSdigital/ras-frontstage/blob/master/frontstage/error_handlers.py#L30).

File: frontstage/controllers/oauth_controller.py (hunk 3 above)
Comment (jcox-dev) [reply frontstage/controllers/oauth_controller.py]: Unexpected errors are handled by the generic [catch all](https://github.com/ONSdigital/ras-frontstage/blob/master/frontstage/error_handlers.py#L45). I'm not sure whether there's an intention to change to a more defensive style but probably worth raising.

File: frontstage/views/sign_in/sign_in.py (hunk 7 above)
Comment (jcox-dev) [reply frontstage/views/sign_in/sign_in.py]: @JamesGardiner Not a fan of an extra `else` but if it reads clunky then I'll add it here.

Comment (JamesGardiner) [reply frontstage/views/sign_in/sign_in.py]: Nope, just a suggestion 👍

**Comment by jcox-dev** [comment]
[ras-frontstage should directly call django service (not via ras-frontstage-api)](https://trello.com/c/luQFWqS0/104-ras-frontstage-should-directly-call-django-service-not-via-ras-frontstage-api)

**Comment by rhiGriff** [comment]
Can you add any config changes made to the confluence page please? https://digitaleq.atlassian.net/wiki/spaces/RASB/pages/406257665/Pending+Release+Config+Changes

File: config.py [hunk 9]
```diff
@@ -30,6 +33,9 @@ class Config(object):
     PASSWORD_MIN_LENGTH = 8
     PASSWORD_MAX_LENGTH = 160
 
+    OAUTH_SERVICE_URL = os.getenv('OAUTH_SERVICE_URL', 'http://localhost:8040')
```
Comment (gemmai95) [inline config.py]: Don't think these should have defaults here only in the dev config.

File: frontstage/views/sign_in/sign_in.py [hunk 10]
```diff
@@ -30,22 +35,17 @@ def login():
...
+        try:
+            oauth2_token = oauth_controller.sign_in(username, password)
+        except OAuth2Error as exc:
+            error_message = exc.message
+            if BAD_AUTH_ERROR in error_message:
                 return render_template('sign-in/sign-in.html', form=form, data={"error": {"type": "failed"}})
-            elif 'User account locked' in error_message:
+            elif ACCOUNT_LOCKED_ERROR in error_message:
                 logger.debug('User account is locked on the OAuth2 server')
                 return render_template('sign-in/sign-in.trouble.html', form=form, data={"error": {"type": "account locked"}})
```
Comment (gemmai95) [inline frontstage/views/sign_in/sign_in.py]: I don't think this is handled by the UI anymore or it shouldn't be we don't want to be feeding back that the account is locked

File: config.py (hunk 9 above)
Comment (jcox-dev) [reply config.py]: @gemmai95 I agree. And would be consistent with the new r-ops changes.

File: frontstage/views/sign_in/sign_in.py (hunk 10 above)
Comment (jcox-dev) [reply frontstage/views/sign_in/sign_in.py]: @gemmai95 can this case be removed then? (Should it be done so in this PR or separately?).

Comment (gemmai95) [reply frontstage/views/sign_in/sign_in.py]: I have removed it in my PR.

Comment (jcox-dev) [reply frontstage/views/sign_in/sign_in.py]: I'll drop it out here as well then.

File: config.py [hunk 11]
```diff
@@ -75,6 +83,16 @@ class DevelopmentConfig(Config):
     SECURITY_USER_PASSWORD = os.getenv('SECURITY_USER_PASSWORD', 'secret')
     BASIC_AUTH = (SECURITY_USER_NAME, SECURITY_USER_PASSWORD)
 
+    OAUTH_URL = os.getenv('OAUTH_URL', 'http://localhost:8040')
+    OAUTH_CLIENT_ID = os.getenv('OAUTH_CLIENT_ID', 'admin')
```
Comment (tej99) [inline config.py]: Looks like this is getting set to ons@ons.gov as default

File: config.py [hunk 12]
```diff
@@ -75,6 +83,16 @@ class DevelopmentConfig(Config):
...
+    OAUTH_CLIENT_SECRET = os.getenv('OAUTH_CLIENT_SECRET', 'secret')
+    OAUTH_BASIC_AUTH = (OAUTH_CLIENT_ID, OAUTH_CLIENT_SECRET)
+
+    PARTY_URL = os.getenv('PARTY_URL', 'http://localhost:8081')
+    PARTY_USERNAME = os.getenv('PARTY_USERNAME', 'admin')
+    PARTY_PASSWORD = os.getenv('PARTY_PASSWORD', 'secret')
```
Comment (tej99) [inline config.py]: looks like this is getting set to password as default

Comment (jcox-dev) [reply config.py]: @tej99 sounds like the party service failing to be consistent rather than an issue here.

Comment (tej99) [reply config.py]: Sorry that was a comment for OAUTH_CLIENT_SECRET

**Comment by jcox-dev** [comment]
Merged into #333.

**Pull request closed without merging**